File: scripts/smoke_persistence.py
#!/usr/bin/env python3
"""
Smoke test for Photune persistence layer:
- Validates auto-save implementation
- Validates canvas persistence (load/save cycle)
- Validates conflict detection
"""

import sys
import json
from pathlib import Path

def validate_auto_save_ts():
    """Check that auto-save.ts exists and has required exports."""
    auto_save_path = Path(__file__).parent.parent / "src" / "shared" / "lib" / "auto-save.ts"
    
    if not auto_save_path.exists():
        print("[FAIL] auto-save.ts does not exist")
        return False
    
    with open(auto_save_path, 'r') as f:
        content = f.read()
    
    required_exports = [
        'export async function autoSaveProject',
        'lastSaveTime',
        'conflict detection',
        'exponential backoff',
    ]
    
    for export in required_exports:
        if export not in content:
            print(f"[WARN] auto-save.ts missing: {export}")
    
    print("[PASS] auto-save.ts exists with required structure")
    return True

def validate_canvas_persistence_ts():
    """Check that canvas-persistence.ts exists and has required exports."""
    persistence_path = Path(__file__).parent.parent / "src" / "features" / "editor" / "lib" / "canvas-persistence.ts"
    
    if not persistence_path.exists():
        print("[FAIL] canvas-persistence.ts does not exist")
        return False
    
    with open(persistence_path, 'r') as f:
        content = f.read()
    
    required_exports = [
        'export function extractCanvasToPersistence',
        'export async function hydrateCanvasFromPersistence',
        'schema validation',
        'corrupted state handling',
    ]
    
    for export in required_exports:
        if export not in content:
            print(f"[WARN] canvas-persistence.ts missing: {export}")
    
    print("[PASS] canvas-persistence.ts exists with required structure")
    return True

def validate_editor_client_integration():
    """Check that EditorClient.tsx imports and integrates auto-save."""
    editor_path = Path(__file__).parent.parent / "src" / "features" / "editor" / "components" / "EditorClient.tsx"
    
    if not editor_path.exists():
        print("[FAIL] EditorClient.tsx does not exist")
        return False
    
    with open(editor_path, 'r') as f:
        content = f.read()
    
    required_patterns = [
        'import { autoSaveProject }',
        'import { extractCanvasToPersistence }',
        'autoSaveTimerRef',
        'setInterval',
        '30_000',
        'lastSaveTime',
    ]
    
    missing = []
    for pattern in required_patterns:
        if pattern not in content:
            missing.append(pattern)
    
    if missing:
        print(f"[WARN] EditorClient.tsx missing patterns: {missing}")
    else:
        print("[PASS] EditorClient.tsx successfully integrated auto-save")
    
    return len(missing) == 0

def validate_package_json_scripts():
    """Check that package.json has persistence smoke test script."""
    pkg_path = Path(__file__).parent.parent / "package.json"
    
    if not pkg_path.exists():
        print("[FAIL] package.json does not exist")
        return False
    
    with open(pkg_path, 'r') as f:
        pkg = json.load(f)
    
    scripts = pkg.get('scripts', {})
    
    if 'smoke:persistence' not in scripts:
        print("[WARN] package.json missing smoke:persistence script")
        return False
    
    if 'smoke:persistence' not in scripts['smoke']:
        print("[WARN] smoke script does not include smoke:persistence")
        return False
    
    print("[PASS] package.json scripts correctly configured")
    return True

def main():
    print("\n========================================")
    print("Photune Persistence Layer Smoke Test")
    print("========================================\n")
    
    tests = [
        ("Auto-save module", validate_auto_save_ts),
        ("Canvas persistence module", validate_canvas_persistence_ts),
        ("EditorClient integration", validate_editor_client_integration),
        ("Package.json scripts", validate_package_json_scripts),
    ]
    
    results = []
    for test_name, test_fn in tests:
        try:
            result = test_fn()
            results.append((test_name, result))
        except Exception as e:
            print(f"[ERROR] {test_name}: {e}")
            results.append((test_name, False))
    
    print("\n========================================")
    print("Test Summary")
    print("========================================")
    
    passed = sum(1 for _, r in results if r)
    total = len(results)
    
    for test_name, result in results:
        status = "✓ PASS" if result else "✗ FAIL"
        print(f"{status}: {test_name}")
    
    print(f"\nTotal: {passed}/{total} tests passed")
    print("========================================\n")
    
    return 0 if passed == total else 1

if __name__ == "__main__":
    sys.exit(main())

File: scripts/smoke-persistence.mjs
#!/usr/bin/env node

/**
 * Smoke Test: Canvas Persistence
 * 
 * Validates:
 * 1. Auto-save creates valid JSON
 * 2. Canvas hydration validates schema
 * 3. Reload cycle preserves state
 */

import fs from 'fs';
import path from 'path';

const ERRORS = [];
const WARNINGS = [];
const CHECKS = [];

function check(name, pass, details = '') {
  const status = pass ? '✓' : '✗';
  const line = `${status} ${name}${details ? ' — ' + details : ''}`;
  CHECKS.push({ name, pass, details });
  console.log(`  ${line}`);
  if (!pass) ERRORS.push(name);
}

function warn(msg) {
  WARNINGS.push(msg);
  console.log(`  ⚠ ${msg}`);
}

console.log('\n🔍 PERSISTENCE SMOKE TESTS\n');

// Test 1: Auto-save.ts exists and exports
console.log('1. Auto-save Module');
try {
  const autoSavePath = 'src/shared/lib/auto-save.ts';
  const exists = fs.existsSync(autoSavePath);
  check('auto-save.ts exists', exists);

  if (exists) {
    const content = fs.readFileSync(autoSavePath, 'utf-8');
    check('exports autoSaveProject', content.includes('export async function autoSaveProject'));
    check('includes collision detection', content.includes('lastKnownUpdateTime'));
    check('includes retry logic in docs', content.includes('exponential backoff'));
  }
} catch (e) {
  ERRORS.push('Failed to check auto-save.ts: ' + e.message);
}

// Test 2: Canvas persistence module
console.log('\n2. Canvas Persistence Module');
try {
  const persistPath = 'src/features/editor/lib/canvas-persistence.ts';
  const exists = fs.existsSync(persistPath);
  check('canvas-persistence.ts exists', exists);

  if (exists) {
    const content = fs.readFileSync(persistPath, 'utf-8');
    check('exports hydrateCanvasFromPersistence', content.includes('export async function hydrateCanvasFromPersistence'));
    check('exports extractCanvasToPersistence', content.includes('export function extractCanvasToPersistence'));
    check('includes JSON validation', content.includes('validateCanvasJson'));
    check('includes error handling', content.includes('catch'));
  }
} catch (e) {
  ERRORS.push('Failed to check canvas-persistence.ts: ' + e.message);
}

// Test 3: EditorClient integration
console.log('\n3. EditorClient Integration');
try {
  const editorPath = 'src/features/editor/components/EditorClient.tsx';
  const content = fs.readFileSync(editorPath, 'utf-8');
  
  check('imports autoSaveProject', content.includes('import { autoSaveProject }'));
  check('imports extractCanvasToPersistence', content.includes('import { extractCanvasToPersistence }'));
  check('includes auto-save timer effect', content.includes('autoSaveTimerRef.current = setInterval'));
  check('includes lastSaveTime state', content.includes('lastSaveTime'));
  check('has collision detection in effect', content.includes('Conflict'));
} catch (e) {
  ERRORS.push('Failed to check EditorClient: ' + e.message);
}

// Test 4: Type safety
console.log('\n4. Type Safety');
try {
  const autoSavePath = 'src/shared/lib/auto-save.ts';
  const content = fs.readFileSync(autoSavePath, 'utf-8');
  
  check('defines AutoSaveConfig type', content.includes('export type AutoSaveConfig'));
  check('defines AutoSaveResult type', content.includes('export type AutoSaveResult'));
  check('includes proper return types', content.includes('Promise<AutoSaveResult>'));
} catch (e) {
  ERRORS.push('Failed to check types: ' + e.message);
}

// Test 5: Error handling
console.log('\n5. Error Handling');
try {
  const autoSavePath = 'src/shared/lib/auto-save.ts';
  const persistPath = 'src/features/editor/lib/canvas-persistence.ts';
  
  const autoSaveContent = fs.readFileSync(autoSavePath, 'utf-8');
  const persistContent = fs.readFileSync(persistPath, 'utf-8');
  
  check('auto-save logs errors', autoSaveContent.includes('logError'));
  check('auto-save logs info', autoSaveContent.includes('logInfo'));
  check('persistence validates before load', persistContent.includes('validateCanvasJson'));
  check('persistence has try-catch blocks', persistContent.match(/catch/g)?.length >= 2);
} catch (e) {
  ERRORS.push('Failed to check error handling: ' + e.message);
}

// Summary
console.log('\n' + '='.repeat(60));
const passCount = CHECKS.filter(c => c.pass).length;
const failCount = CHECKS.filter(c => !c.pass).length;
const total = CHECKS.length;

console.log(`\nResults: ${passCount}/${total} checks passed`);

if (failCount > 0) {
  console.log(`\n❌ FAILED CHECKS (${failCount}):`);
  ERRORS.forEach(e => console.log(`   - ${e}`));
  process.exit(1);
}

if (WARNINGS.length > 0) {
  console.log(`\n⚠ WARNINGS (${WARNINGS.length}):`);
  WARNINGS.forEach(w => console.log(`   - ${w}`));
}

console.log('\n✅ Persistence tests passed!\n');
process.exit(0);
